tools/network/dnssec/resolver.go

```diff
@@ -0,0 +1,331 @@
+)
+
+const defaultMaxHops = 10
+
+var defaultDnssecAwareNSServers = []string{"1.1.1.1:53", "8.8.8.8:53"}
+
+// Resolver provides DNSSEC resolution
+type Resolver struct {
+	resolver   netResolverIf
+	trustChain *trustChain
+	maxHops    int
+}
+
+// DefaultResolver with one DNS server and 1 second timeout
+var DefaultResolver = MakeDnssecResolver(defaultDnssecAwareNSServers, time.Second)
+
+// MakeDnssecResolver return resolver from given NS servers and timeout duration
+func MakeDnssecResolver(servers []string, timeout time.Duration) (r Resolver) {
+	rs := &resolverImpl{readTimeout: timeout, servers: servers, rootAnchor: rootAnchorXML}
+
+	if len(rs.servers) == 0 {
+		rs.servers = append(rs.servers, defaultDnssecAwareNSServers...)
+	}
+	r.resolver = rs
+	r.trustChain = makeTrustChain(r.resolver)
+	r.maxHops = defaultMaxHops
+	return
+}
+
+type netResolverIf interface {
+	query(ctx context.Context, domain string, qtype uint16) (resp *dns.Msg, err error)
+	queryRRSet(ctx context.Context, domain string, qtype uint16) ([]dns.RR, []dns.RRSIG, error)
+	rootTrustAnchor() ([]dns.DS, error)
+
+	// test functions
+	serverList() []string
+}
+
+// TLSARec represents TLSA record content
+type TLSARec struct {
+	Usage        uint8
+	Selector     uint8
+	MatchingType uint8
+	Certificate  string `dns:"hex"`
+}
+
+// lookupImpl makes DNS request for a zone and verifies response signature
+func (r *Resolver) lookupImpl(ctx context.Context, name string, qt uint16) (rrSet []dns.RR, err error) {
+	rrSet, rrSig, err := r.resolver.queryRRSet(ctx, name, qt)
+	if err != nil {
+		return
+	}
+	err = r.trustChain.authenticate(ctx, rrSet, rrSig)
+	return
+}
+
+// lookupImplCnameAware like lookupImpl makes DNS request for a zone and verifies response signature
+// but also it is aware about possible A/CNAME entries for A-requests and supposed to be used for CNAME alias resolution.
+// if CNAME signature presents for requested domain name, then consider the response as CNAME
+// and extract CNAME record(s) and its RRSIG
+func (r *Resolver) lookupImplCnameAware(ctx context.Context, name string, qt uint16) (rrSet []dns.RR, err error) {
+	rrSet, rrSig, err := r.resolver.queryRRSet(ctx, name, qt)
+	if err != nil {
+		return
+	}
+
+	// As https://tools.ietf.org/html/rfc1034 section 3.6.2 says
+	// NS can return CNAME for A request if there no A but CNAME
+	// in real world such A-response can also have second-level CNAME alias and A records with signatures
+	//
+	// so that leave only CNAME RR matching to requested name to verify signature and return the filtered set
+	if qt == dns.TypeA || qt == dns.TypeAAAA {
+		newRRSig := make([]dns.RRSIG, 0, len(rrSig))
+		for _, sig := range rrSig {
+			if sig.Hdr.Name == name {
+				if sig.TypeCovered == dns.TypeCNAME || sig.TypeCovered == dns.TypeA || sig.TypeCovered == dns.TypeAAAA {
+					if len(newRRSig) == 0 {
+						newRRSig = append(newRRSig, sig)
+					} else if sig.TypeCovered == newRRSig[len(newRRSig)-1].TypeCovered {
```

Comment:
Is it okay to ignore if TypeCovered is different? 
Also, simpler to say: 
sig.TypeCovered == newRRSig[0].TypeCovered

Reply:
This is a special case for following CNAMEs. The response might contain multiple CNAMEs with RRSIGS for different names in the alias chain, plus it also can contain A and its RRSIG (if in cache)
The goal is to check only one CNAME at the time and then find A and RRSIG by ourselves.
Yes, this can be optimized to gather all the info in the response and use it... Maybe in next PR? Trade-off is the code simplicity vs DNS requests emitted.